### PR TITLE
Add parameter for FW negative cycle detection

### PR DESCRIPTION
**Pull request by kennyballou**
In b1f9e1e7e5 ("Implemented heuristics for Floyd-Warshall", Sat Jul 14
12:16:16 2018 +0300), some heuristics were added to improve the
performance of the Floyd-Warshall algorithm.  However, as an unintended
consequence, the FW algorithm could no longer be used to detect negative
cycles.

Add a parameter to the algorithm constructor, maintaining the existing
behavior, which enables users to detect negative cycle detection.
Furthermore, add a method to check for negative cycles.

This resolves [#1096](https://github.com/jgrapht/jgrapht/issues/1096).

----

- [X] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [X] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [X] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [X] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [X] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [X] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [X] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/FloydWarshallShortestPaths.java [hunk 1]
```diff
@@ -305,11 +340,11 @@ private void lazyCalculateMatrix()
         // run fw alg
         for (int k = minDegreeTwo; k < n; k++) {
             for (int i = minDegreeOne; i < n; i++) {
-                if (i == k) {
+                if (!detectNegativeCycles && i == k) {
```
[the remark this answers is not in the document]
Comment (kennyballou) [reply jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/FloydWarshallShortestPaths.java:343]: Normally, when a graph does not contain a strictly negative cycle, no path through the graph will be "closer" than a self-loop on a node.  This is the basis of the heuristic of skipping specific iterations.  However, when a negative cycle is possible or a negative cycle does appear in the graph, there is a path that is not the self-loop that is "closer" than the self loop.  The Floyd-Warshall algorithm can detect these negative cycles (when performed without the optimizing heuristic) and they appear in the diagonal of the weighted adjacency matrix.

I'm going to amend the PR with another test case for non-negative cycle graphs.

**Comment by jkinable** [comment]
I've blocked some time in my calendar on Friday to work on my backlog of reviews...
